File: scripts/unittests.jl
using Base.Filesystem
using Test
using ModelicaScriptingTools: withOMC, testmodel, loadModel
using OMJulia: sendExpression

moroot = dirname(@__DIR__)
outdir = joinpath(moroot, "out")
if !ispath(outdir)
    mkdir(outdir)
end
refdir = joinpath(moroot, "regRefData")

# relative tolerance for unit tests
# This is chosen a little larger than the tolerance used for most simulations
# (1e-6) to allow for some error propagation.
# Models that do not have a controlled timing of AP events (namely experiments
# including the N cell model) may still fail tests after small and harmless
# changes, because the error accumulates over time.
rrtol = 1e-5
withOMC(outdir, moroot) do omc
    sendExpression(omc, "setCommandLineOptions(\"-d=newInst,nfAPI\")")
    @testset "Simulate examples" begin
        @testset "SodiumChannelSteady" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.SodiumChannelSteady"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "SodiumChannelIV" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.SodiumChannelIV"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "InwardRectifierLin" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.InwardRectifierLin"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "GHKFlux" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.GHKFlux"; override=Dict(
                "stopTime"=>100, "numberOfIntervals"=>1000
            ), refdir=refdir, regRelTol=rrtol)
        end
        @testset "LTypeCalciumSteady" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.LTypeCalciumSteady"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "LTypeCalciumIV" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.LTypeCalciumIV"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "LTypeCalciumIVN" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.LTypeCalciumIVN"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "LTypeCalciumStep" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.LTypeCalciumStep"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "TransientOutwardSteady" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.TransientOutwardSteady"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "TransientOutwardIV" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.TransientOutwardIV"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "RapidDelayedRectifierSteady" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.RapidDelayedRectifierSteady"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "RapidDelayedRectifierIV" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.RapidDelayedRectifierIV"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "HyperpolarizationActivatedSteady" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.HyperpolarizationActivatedSteady"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "HyperpolarizationActivatedIV" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.HyperpolarizationActivatedIV"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "SustainedInwardSteady" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.SustainedInwardSteady"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "SustainedInwardIV" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.SustainedInwardIV"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "SustainedInwardIVKurata" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.SustainedInwardIVKurata"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "SodiumPotassiumPumpLin" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.SodiumPotassiumPumpLin"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "SodiumCalciumExchangerRampInada" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.SodiumCalciumExchangerRampInada"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "SodiumCalciumExchangerLinMatsuoka" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.SodiumCalciumExchangerLinMatsuoka"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "SodiumCalciumExchangerLinKurata" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.SodiumCalciumExchangerLinKurata"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "AllCells" begin
            # we use a different tolerance here, because errors in N cell model accumulate over time
            testmodel(omc, "InaMo.Examples.FullCell.AllCells"; refdir=refdir, regRelTol=1e-3)
        end
        @testset "AllCellsC" begin
            # we use a different tolerance here, because errors in N cell model accumulate over time
            testmodel(omc, "InaMo.Examples.FullCell.AllCellsC"; refdir=refdir, regRelTol=1e-3)
        end
        @testset "FullCellSpon" begin
            testmodel(omc, "InaMo.Examples.FullCell.FullCellSpon"; override=Dict(
                "stopTime"=>0.5, "numberOfIntervals"=>5000, "variableFilter"=>raw"cell\.(naca|cal)\.i",
                "tolerance"=>1e-6
            ), refdir=refdir, regRelTol=rrtol)
        end
        @testset "CaHandlingApprox" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.CaHandlingApprox"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "CaBuffer" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.CaBuffer"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "CaBuffer2" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.CaBuffer2"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "CaDiffusion" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.CaDiffusion"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "CaSERCA" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.CaSERCA"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "CaRyanodineReceptor" begin
            testmodel(omc, "InaMo.Examples.ComponentTests.CaRyanodineReceptor"; refdir=refdir, regRelTol=rrtol)
        end
        @testset "AtrialCellSmokeTest" begin
            # only test if model can be loaded, simulation not working yet
            @test isnothing(loadModel(omc, "InaMo.Examples.FullCell.AllCells"))
            # testmodel(omc, "InaMo.Examples.FullCell.AllCells"; refdir=refdir, regRelTol=rrtol)
        end
        #=
        @testset "SteadyStates" begin
            # only test if model can be simulated
            # TODO this currently does not work, because internal functions cannot be referenced from outside a component
            testmodel(omc, "InaMo.Examples.ComponentTests.SteadyStates"; override=Dict(
                "stopTime"=>0.01
            ), refdir=nothing, regRelTol=rrtol)
        end
        =#
    end
end

File: scripts/unittests.py
import unittest
from pathlib import Path
import OMPython
import os
import time


def escape_mostring(s):
    t = str.maketrans({
        "\\": "\\\\",
        '"': '\\"',
        '?': '\\?',
        '\a': '\\a',
        '\b': '\\b',
        '\f': '\\f',
        '\n': '\\n',
        '\r': '\\r',
        '\t': '\\t',
        '\v': '\\v',
    })
    return str(s).translate(t)


def assert_sim_noerror(test, model, override=None):
    if override is None:
        override = {}
    omc = TestIonChannels.omc
    r = omc.sendExpression("loadModel({})".format(model))
    test.assertTrue(r)
    es = omc.sendExpression("getErrorString()")
    test.assertEqual(0, len(es), msg=es)
    start, stop, tol, noi, _ = omc.sendExpression(
        "getSimulationOptions({})".format(model)
    )
    settings = {
        "startTime": start, "stopTime": stop,
        "tolerance": tol, "numberOfIntervals": noi,
        "outputFormat": '"csv"'
    }
    for x in settings:
        if x in override:
            settings[x] = override[x]
    r = omc.sendExpression(
        "simulate({}, {})".format(
            model,
            ", ".join(["{}={}".format(x, settings[x]) for x in settings])
        )
    )
    test.assertFalse("| warning |" in r["messages"], msg=r["messages"])
    test.assertFalse(
        r["messages"].startswith("Simulation execution failed"),
        msg=r["messages"]
    )
    es = omc.sendExpression("getErrorString()")
    test.assertEqual(0, len(es), msg=es)


class TestIonChannels(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        moroot = Path(__file__).parent.parent
        outdir = moroot / "out"
        os.chdir(outdir)
        if not outdir.is_dir():
            outdir.mkdir()
        cls.omc = OMPython.OMCSessionZMQ()
        mopath = cls.omc.sendExpression("getModelicaPath()")
        mopath = ":".join([mopath, escape_mostring(moroot)])
        print("Setting MODELICAPATH to ", mopath)
        cls.omc.sendExpression('setModelicaPath(\"{}\")'.format(mopath))
        # load Modelica standard library
        cls.omc.sendExpression("loadModel(Modelica)")

    @classmethod
    def tearDownClass(cls):
        print("Closing OMC session")
        time.sleep(1)
        cls.omc.sendExpression("quit()", parsed=False)
        print("Done")

    def test_sodium_channel_steady(self):
        assert_sim_noerror(self, "InaMo.Examples.SodiumChannelSteady")

    def test_SodiumChannelIV(self):
        assert_sim_noerror(self, "InaMo.Examples.SodiumChannelIV")

    def test_InwardRectifierLin(self):
        assert_sim_noerror(self, "InaMo.Examples.InwardRectifierLin")

    def test_GHKFlux(self):
        assert_sim_noerror(self, "InaMo.Examples.GHKFlux", override={
            "stopTime": 100, "numberOfIntervals": 1000
        })

    def test_LTypeCalciumSteady(self):
        assert_sim_noerror(self, "InaMo.Examples.LTypeCalciumSteady")

    def test_LTypeCalciumIV(self):
        assert_sim_noerror(self, "InaMo.Examples.LTypeCalciumIV")

    def test_LTypeCalciumIVN(self):
        assert_sim_noerror(self, "InaMo.Examples.LTypeCalciumIVN")

    def test_LTypeCalciumStep(self):
        assert_sim_noerror(self, "InaMo.Examples.LTypeCalciumStep")

    def test_TransientOutwardSteady(self):
        assert_sim_noerror(self, "InaMo.Examples.TransientOutwardSteady")

    def test_TransientOutwardIV(self):
        assert_sim_noerror(self, "InaMo.Examples.TransientOutwardIV")

    def test_RapidDelayedRectifierSteady(self):
        assert_sim_noerror(self, "InaMo.Examples.RapidDelayedRectifierSteady")

    def test_RapidDelayedRectifierIV(self):
        assert_sim_noerror(self, "InaMo.Examples.RapidDelayedRectifierIV")

    def test_HyperpolarizationActivatedSteady(self):
        assert_sim_noerror(
            self, "InaMo.Examples.HyperpolarizationActivatedSteady"
        )

    def test_HyperpolarizationActivatedIV(self):
        assert_sim_noerror(self, "InaMo.Examples.HyperpolarizationActivatedIV")

    def test_SustainedInwardSteady(self):
        assert_sim_noerror(self, "InaMo.Examples.SustainedInwardSteady")

    def test_SustainedInwardIV(self):
        assert_sim_noerror(self, "InaMo.Examples.SustainedInwardIV")

    def test_SustainedInwardIVKurata(self):
        assert_sim_noerror(self, "InaMo.Examples.SustainedInwardIVKurata")

    def test_SodiumPotassiumPumpLin(self):
        assert_sim_noerror(self, "InaMo.Examples.SodiumPotassiumPumpLin")

    def test_SodiumCalciumExchangerRamp(self):
        assert_sim_noerror(self, "InaMo.Examples.SodiumCalciumExchangerRamp")

    def test_SodiumCalciumExchangerLin(self):
        assert_sim_noerror(self, "InaMo.Examples.SodiumCalciumExchangerLin")


if __name__ == '__main__':
    unittest.main()
